### PR TITLE
wave3: Multi-Scale Attention with Pooled Context (cross-dataset)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Standard attention operates at a single resolution. In CFD point clouds, there are important features at multiple scales: local surface curvature (few points), pressure gradients across panels (many points), and global flow regime (all points). Multi-scale attention, where each head also attends to pooled representations of larger neighborhood contexts, should improve the model's ability to capture multi-resolution flow physics without increasing the point count.

## Implementation

Add to `train.py`:

```python
class MultiScaleAttention(nn.Module):
    """
    Multi-scale attention: standard attention + pooled context attention.
    Each transformer block gets both local (all-to-all) and global (pooled) context.
    """
    def __init__(self, dim, num_heads, pool_sizes=(4, 16), dropout=0.0):
        super().__init__()
        self.num_heads = num_heads
        self.head_dim = dim // num_heads
        self.scale = self.head_dim ** -0.5
        
        # Standard QKV projection
        self.qkv = nn.Linear(dim, 3 * dim, bias=False)
        self.proj = nn.Linear(dim, dim, bias=False)
        
        # Pooled context projections (one per pool scale)
        self.pool_sizes = pool_sizes
        self.pool_projs = nn.ModuleList([
            nn.Linear(dim, dim // len(pool_sizes), bias=False)
            for _ in pool_sizes
        ])
        self.pool_out_proj = nn.Linear(dim, dim, bias=False)
        self.norm_pool = nn.LayerNorm(dim)
    
    def forward(self, x):
        B, N, D = x.shape
        
        # Standard attention
        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, self.head_dim).permute(2, 0, 3, 1, 4)
        q, k, v = qkv.unbind(0)
        attn = (q @ k.transpose(-2, -1)) * self.scale
        attn = F.softmax(attn, dim=-1)
        x_attn = (attn @ v).transpose(1, 2).reshape(B, N, D)
        
        # Pooled context: for each pool_size, avg-pool then project
        pool_contexts = []
        for pool_size, pool_proj in zip(self.pool_sizes, self.pool_projs):
            # Adaptive avg pool along N dimension
            x_pool = x.transpose(1, 2)  # [B, D, N]
            x_pool = F.adaptive_avg_pool1d(x_pool, max(1, N // pool_size))  # [B, D, N//pool_size]
            x_pool = x_pool.transpose(1, 2)  # [B, N//pool_size, D]
            pool_contexts.append(pool_proj(x_pool))  # [B, N//pool_size, D//num_scales]
        
        # Concatenate pooled contexts and broadcast back
        pool_context = torch.cat(pool_contexts, dim=-1)  # [B, N//min_pool, D]
        # Simple: use mean pooled context as additive bias
        global_ctx = pool_context.mean(dim=1, keepdim=True)  # [B, 1, D]
        x_pool_out = self.pool_out_proj(self.norm_pool(global_ctx))  # [B, 1, D]
        
        return self.proj(x_attn) + x_pool_out.expand(B, N, D)
```

Add CLI argument:
```python
parser.add_argument('--multi-scale-attn', action='store_true',
    help='Use multi-scale attention with pooled context at scales 4x and 16x')
parser.add_argument('--multi-scale-pool-sizes', type=str, default='4,16',
    help='Comma-separated pool reduction factors for multi-scale attention')
```

Use `--wandb_group wave3-multi-scale-attention` for grouping.

### TandemFoil
```bash
python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --multi-scale-attn --multi-scale-pool-sizes 4,16 \
  --epochs 999 --wandb_group wave3-multi-scale-attention
```

### TandemFoil Paper
```bash
python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --multi-scale-attn --multi-scale-pool-sizes 4,16 \
  --epochs 999 --wandb_group wave3-multi-scale-attention
```

### AirfRANS
```bash
python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --multi-scale-attn --multi-scale-pool-sizes 4,16 \
  --epochs 999 --wandb_group wave3-multi-scale-attention
```

### DrivAerML
```bash
SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --multi-scale-attn --multi-scale-pool-sizes 4,16 \
  --wandb_group wave3-multi-scale-attention
```

## Current Baselines

| Dataset | Metric | Best Value | PR |
|---------|--------|-----------|-----|
| TandemFoil | val_primary/surface_pressure_mae | 26.06 | #2887 |
| TandemFoil Paper | val_primary/field_mse | TBD | #2947/2948/2949 |
| AirfRANS | val_primary/surface_mse | 0.000627 | #2902 |
| DrivAerML | val_primary/surface_rel_l2_pct | 3.997% | #2898 |

## Expected Outcome

Multi-scale attention allows the model to simultaneously reason about local geometric features and global flow patterns. In CFD, this is critical: local surface curvature affects boundary layer attachment while global geometry determines wake topology. The pooled context adds negligible compute overhead. Expected: 2-4% improvement especially on DrivAerML where multi-scale geometric features are richest.

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
